### PR TITLE
[Enhancement] Refactor the create table logic to reduce the I/O usage.

### PR DESCRIPTION
**Pull request by chaoyli**
Currently, the create table will be splited multiple create_tablet tasks. Every create_tablet task will need two I/O operations, and always hold the lock, per RPC per tablet.
The procedure is as followes.
```
1. hold the tablet_shard lock      (hold the lock always)
2. create the tablet dir           (First I/O)
3. create tablet meta              (Second I/O)
4. add the tablet to the tablet_map
5. report the status to FE
```

If the concurrency of create table is high, the number of create_tablet tasks may be very high. It will be stuck on the disk I/O and the lock. So I batch the I/O operation for table granularity. Every table will cost constant four I/O operations. As well the I/O opeartions is moved out of the lock. Finally, it only have one RPC to response the status. The new operation is as follows.
```
1. generate tablet dirs
2. save_create_txn(TXN_PREPARED)  (First I/O)
3. mkdir tablet dirs
4. sync tablet dirs               (Second I/O)
5. create tablet meta             (Third I/O)
6. save_create_txn(TXN_COMMITTED) (Four I/O)
7. hold the tablet_shard lock     (hold lock only for a while)
8. add the tablet to the tablet_map
9. report the status to FE
```

To complete the optimization, FE/BE use transaction mechanism to guarantee the exactly-once for create table.
The transaction is only one-phase, not two-phases. When the BE succeed to create the table, the tablets will be added to the map. So we need to remove the tablet when FE failed, partial BEs succeed.
```
1. FE send an abort request to BEs at best effort.
2. BE remove the timeout transactions through an daemon thread.
3. BE report the tablets list to FE, FE send an request to remove the extra tablets.
```

If the transaction failed or timeout, the FE side remove the transaction and send abort request to BE. The BE side will remove the

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [x] Yes, this PR will result in a change in behavior.
- [ ] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [x] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:

- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.2
  - [x] 3.1
  - [x] 3.0
  - [ ] 2.5

### CODE REVIEW

**Comment by chaoyli** [comment]
I have tested with 1000 concurrency to create the table. The latency is reduced from 3min20s to 20s.
```
CREATE TABLE test$i (k1 int, k2 int) ENGINE=OLAP DUPLICATE KEY(k1, k2) DISTRIBUTED BY HASH(k1) BUCKETS 100 PROPERTIES ('replication_num' = '1')
```

File: gensrc/thrift/AgentService.thrift [hunk 1]
```diff
@@ -116,6 +116,22 @@ struct TCreateTabletReq {
     20: optional bool create_schema_file = true;
 }
 
+struct TCreateTableReq {
+    1: optional list<TCreateTabletReq> create_tablet_reqs;
+    2: optional i32 timeout = 0; // unit: second
+}
+
+struct TAbortTxnReq {
+    1: required Types.TTransactionId txn_id;
```
Comment (imay) [inline gensrc/thrift/AgentService.thrift:125]: All fields in thrift/protobuf should be optional.

File: gensrc/thrift/AgentService.thrift [hunk 2]
```diff
@@ -116,6 +116,22 @@ struct TCreateTabletReq {
...
+}
+
+struct TReq {
```
Comment (imay) [inline gensrc/thrift/AgentService.thrift:128]: This name is too short, need to be more precise.

File: fe/fe-core/src/main/java/com/starrocks/common/Config.java [hunk 3]
```diff
@@ -760,6 +760,12 @@ public class Config extends ConfigBase {
     @ConfField(mutable = true)
     public static int tablet_create_timeout_second = 10;
 
+    @ConfField(mutable = true)
+    public static int create_table_timeout_second = 60;
```
Comment (nshangyiming) [inline fe/fe-core/src/main/java/com/starrocks/common/Config.java:764]: plz add some comments about these two configurations.

File: gensrc/thrift/AgentService.thrift (hunk 1 above)
Comment (chaoyli) [reply gensrc/thrift/AgentService.thrift:125]: Fixed

File: fe/fe-core/src/main/java/com/starrocks/common/Config.java (hunk 3 above)
Comment (chaoyli) [reply fe/fe-core/src/main/java/com/starrocks/common/Config.java:764]: fixed

**Comment by nshangyiming** [comment]
@mergify rebase

File: gensrc/thrift/AgentService.thrift [hunk 4]
```diff
@@ -122,6 +122,22 @@ struct TCreateTabletReq {
...
+struct TCreateTableReq {
+    1: optional list<TCreateTabletReq> create_tablet_reqs;
+    2: optional i32 timeout = 0; // unit: second
+}
+
+struct TAbortTxnReq {
+    1: optional Types.TTransactionId txn_id;
+}
+
+struct TReq {
```
Comment (nshangyiming) [inline gensrc/thrift/AgentService.thrift:134]:
```suggestion
struct TTransactionalReq {
```
to differentiate with original task framework

File: gensrc/thrift/AgentService.thrift [hunk 5]
```diff
@@ -122,6 +122,22 @@ struct TCreateTabletReq {
     20: optional bool create_schema_file = true;
 }
 
+struct TCreateTableReq {
```
Comment (nshangyiming) [inline gensrc/thrift/AgentService.thrift:125]:
```suggestion
struct TTxnCreateTableReq {
```
all new req types based on new transactional req framework should start with `TTxn`

File: gensrc/thrift/AgentService.thrift [hunk 6]
```diff
@@ -122,6 +122,22 @@ struct TCreateTabletReq {
...
+    1: optional list<TCreateTabletReq> create_tablet_reqs;
+    2: optional i32 timeout = 0; // unit: second
+}
+
+struct TAbortTxnReq {
```
Comment (nshangyiming) [inline gensrc/thrift/AgentService.thrift:130]:
```suggestion
struct TAbortTransactionalReq {
```

File: gensrc/thrift/AgentService.thrift (hunk 4 above)
Comment (nshangyiming) [inline gensrc/thrift/AgentService.thrift:134]: After offline discussion, should add comments to clarify the necessity of adding a new req framework, including,
1. Why the original task framework cannot work for this situation.
2. What does the txn_id mean, besides for creating table and other ddl operations, is it compatible with data loading process and can be reused by it.

File: fe/fe-core/src/test/java/com/starrocks/common/GenericPoolTest.java [hunk 7]
```diff
@@ -143,6 +144,11 @@ public TAgentResult submit_tasks(List<TAgentTaskRequest> tasks) throws TExceptio
             return null;
         }
 
+        @Override
+        public TAgentResult submit_req(TReq req) throws TException {
```
Comment (nshangyiming) [inline fe/fe-core/src/test/java/com/starrocks/common/GenericPoolTest.java:148]: we don't use this style in java, you can refactor this file

**Review by nshangyiming — COMMENTED** [review]
to be continued

**Comment by nshangyiming** [comment]
@mergify rebase

File: fe/fe-core/src/main/java/com/starrocks/server/LocalMetastore.java [hunk 8]
```diff
@@ -2043,6 +2076,64 @@ private List<CreateReplicaTask> buildCreateReplicaTasks(long dbId, OlapTable tab
...
+                                table.getCompressionType(), indexMeta.getSortKeyIdxes(),
+                                indexMeta.getSortKeyUniqueIds());
+                        CreateTableTask createTableTask = tasks.get(replica.getBackendId());
+                        if (createTableTask == null) {
+                            createTableTask = new CreateTableTask();
+                            long backendId = replica.getBackendId();
+                            createTableTask.setBackendId(backendId);
+                            createTableTask.setTxnId(txnId);
+                            tasks.put(backendId, createTableTask);
+                            LOG.info("backendId: {}", backendId);
```
Comment (nshangyiming) [inline fe/fe-core/src/main/java/com/starrocks/server/LocalMetastore.java:2122]: logging

File: fe/fe-core/src/main/java/com/starrocks/server/LocalMetastore.java [hunk 9]
```diff
@@ -1892,6 +1895,11 @@ void buildPartitions(Database db, OlapTable table, List<PhysicalPartition> parti
         if (partitions.isEmpty()) {
             return;
         }
+
+        long txnId = GlobalStateMgr.getCurrentGlobalTransactionMgr().getTransactionIDGenerator().getNextTransactionId();
+        CreateTableTxnState txnState = new CreateTableTxnState(txnId, Config.create_table_timeout_second);
+        GlobalStateMgr.getCurrentGlobalTransactionMgr().addCreateTableTxn(txnId, txnState);
```
Comment (nshangyiming) [inline fe/fe-core/src/main/java/com/starrocks/server/LocalMetastore.java:1901]: put it in `if (Config.enable_create_table_txn) {...}` ?

File: fe/fe-core/src/main/java/com/starrocks/leader/LeaderImpl.java [hunk 10]
```diff
@@ -357,6 +360,86 @@ public TMasterResult finishTask(TFinishTaskRequest request) {
...
+            if (cn == null) {
+                tStatus.setStatus_code(TStatusCode.CANCELLED);
+                List<String> errorMsgs = new ArrayList<>();
+                errorMsgs.add("backend not exist.");
+                tStatus.setError_msgs(errorMsgs);
+                LOG.warn("backend does not found. host: {}, be port: {}. task: {}", host, bePort, request.toString());
+                return result;
+            }
+        }
+        long backendId = cn.getId();
```
Comment (nshangyiming) [inline fe/fe-core/src/main/java/com/starrocks/leader/LeaderImpl.java:406]: duplicate code with `finishTask()` from L366 - L406, plz refactor

**Pull request closed without merging**